File: fe1-web/Model/Network/Method/Message/data/lao/stateLao.ts
import { Hash } from "Model/Objects/hash";
import { PublicKey } from "Model/Objects/publicKey";
import { Timestamp } from "Model/Objects/timestamp";
import { ActionType, MessageData, ObjectType } from "../messageData";
import { WitnessSignature } from "Model/Objects/witnessSignature";

export class StateLao implements MessageData {

    public readonly object: ObjectType = ObjectType.LAO;
    public readonly action: ActionType = ActionType.STATE;

    public readonly id: Hash;
    public readonly name: string;
    public readonly creation: Timestamp;
    public readonly last_modified: Timestamp;
    public readonly organizer: PublicKey;
    public readonly witnesses: PublicKey[];
    public readonly modification_id: Hash;
    public readonly modification_signatures: WitnessSignature[];

    constructor(msg: Partial<StateLao>) {
        Object.assign(this, msg);
    }

    public verify(): boolean {
        // to be implemented...
        return true;
    }
}

File: be2-scala/src/main/scala/ch/epfl/pop/model/network/method/message/data/lao/StateLao.scala
package ch.epfl.pop.model.network.method.message.data.lao

import ch.epfl.pop.json.MessageDataProtocol._
import ch.epfl.pop.model.network.Parsable
import ch.epfl.pop.model.network.method.message.data.{ActionType, MessageData, ObjectType}
import ch.epfl.pop.model.objects.{Hash, PublicKey, Timestamp, WitnessSignaturePair}
import spray.json._

final case class StateLao(
    id: Hash,
    name: String,
    creation: Timestamp,
    last_modified: Timestamp,
    organizer: PublicKey,
    witnesses: List[PublicKey],
    modification_id: Hash,
    modification_signatures: List[WitnessSignaturePair]
) extends MessageData {
  override val _object: ObjectType = ObjectType.lao
  override val action: ActionType = ActionType.state
}

object StateLao extends Parsable {
  def apply(
      id: Hash,
      name: String,
      creation: Timestamp,
      last_modified: Timestamp,
      organizer: PublicKey,
      witnesses: List[PublicKey],
      modification_id: Hash,
      modification_signatures: List[WitnessSignaturePair]
  ): StateLao = {
    new StateLao(id, name, creation, last_modified, organizer, witnesses, modification_id, modification_signatures)
  }

  override def buildFromJson(payload: String): StateLao = payload.parseJson.asJsObject.convertTo[StateLao]
}
